### ISSUE TITLE
Mismatch in Logits and Labels

### ISSUE THREAD

**Issue by kolloldas** [tensorflow/tensor2tensor#79]
Hi,

The trainer doesn't seem to handle the case properly when the target sequence length is longer than the logits sequence length. I found this out while running the trainer on **wmt_parsing_characters** with **neural_gpu** as the model (for a spell checking task). The _softmax_cross_entropy_with_logits()_ function fails with a shape mismatch error:

`InvalidArgumentError (see above for traceback): logits and labels must be same size: logits_size=[966,256] labels_size=[1008,256]`

The expanded logit shape is **(42, 23, 1, 256)** while the label shape is **(42, 24, 1, 256)**. Since the label/target sequence is longer the _padded_cross_entropy()_ function in common_layers doesn't handle this case:

```
def padded_cross_entropy(logits,
                         labels,
                         label_smoothing,
                         weights_fn=weights_nonzero,
                         reduce_sum=True):
...
with tf.name_scope("padded_cross_entropy", [logits, labels]):
    pad_labels = pad_with_zeros(logits, labels)
    xent = smoothing_cross_entropy(logits, pad_labels, vocab_size, confidence)
...
```

I have fixed the issue by letting _pad_with_zeros()_ return logits also based on a flag:

```
def pad_with_zeros(logits, labels, return_logits=False):
  """Pad labels on the length dimension to match logits length."""
  with tf.name_scope("pad_with_zeros", [logits, labels]):
    logits, labels = pad_to_same_length(logits, labels)
    if len(labels.shape.as_list()) == 3:  # 2-d labels.
      logits, labels = pad_to_same_length(logits, labels, axis=2)
    if return_logits:
      return logits, labels
    else: 
      return labels
```

The updated _padded_cross_entropy()_ function is as follows:

```
def padded_cross_entropy(logits,
                         labels,
                         label_smoothing,
                         weights_fn=weights_nonzero,
                         reduce_sum=True):
...
with tf.name_scope("padded_cross_entropy", [logits, labels]):
    pad_logits, pad_labels = pad_with_zeros(logits, labels, return_logits=True)
    xent = smoothing_cross_entropy(pad_logits, pad_labels, vocab_size, confidence)
...
```

Please let me know if this fix can have any side effects

Thanks!

**Comment by lukaszkaiser** [comment]
I think the correction you present it ok. But I'm wondering how comes this wasn't a problem before. It might be that some other bug recently slipped through. Or maybe not? Thinking...

**Comment by kolloldas** [comment]
_pad_with_zeros()_ is also used in _metrics.py_ by the following functions:

- _padded_accuracy_topk()_
- _padded_sequence_accuracy()_
- _padded_accuracy()_

Wonder if these need the change too..

**Comment by lukaszkaiser** [comment]
This is hopefully corrected in 1.0.11, thanks @kolloldas for the PR! Please give it a try. I'm closing for now, please reopen if you still see the issue.
